src/Group.jl

```diff
@@ -0,0 +1,397 @@
+
+struct LeftInvariantCanonicalMetric <: RiemannianMetric end
+struct RightInvariantCanonicalMetric <: RiemannianMetric end
+
+@traitimpl HasMetric{GroupManifold,LeftInvariantCanonicalMetric}
+
+"""
+    IsMatrixGroup
+
+A `Trait` to indicate that a group's default representation is a matrix
+representation. For groups with a `MultiplicationOperation`, this indicates
+that the group is a subgroup of the General Linear Group.
+"""
+@traitdef IsMatrixGroup{G}
+
+@doc doc"""
+    inv(G::AbstractGroupManifold, x)
+
+Inverse $x^{-1}$ of an element $x$, such that
+$x \cdot x^{-1} = x^{-1} \cdot x = e$.
+"""
+function inv(G::AbstractGroupManifold, x)
+    error("inv not implemented on $(typeof(G)) for point $(typeof(x))")
+end
+
+@doc doc"""
+    identity(G::AbstractGroupManifold, x)
+
+Identity element $e$, such that for any element $x$,
+$x \cdot e = e \cdot x = x$. The returned element is of a similar type to `x`.
+"""
+function identity(G::AbstractGroupManifold, x)
+    error("identity not implemented on $(typeof(G)) for point $(typeof(x))")
+end
+
+@doc doc"""
+    left_action(G::AbstractGroupManifold, x, p)
+
+For an element $p$ of some set, compute the left action of group element $x$
+on $p$, i.e. $x \cdot p$.
+"""
+function left_action(G::AbstractGroupManifold, x, p)
+    error("left_action not implemented on $(typeof(G)) for point $(typeof(x)) and object $(typeof(p))")
+end
+
+@doc doc"""
+    right_action(G::AbstractGroupManifold, x, p)
+
+For an element $p$ of some set, compute the right action of group element $x$
+on $p$, i.e. $p \cdot x$. The default right action is constructed from the left
+action, i.e. $p \cdot x = x^{-1} \cdot p$.
+"""
+function right_action(G::AbstractGroupManifold, x, p)
+    return left_action(G, inv(G, x), p)
+end
+
+"""
+    compose(G::AbstractGroupManifold, x...)
+
+Compose elements of `G` using their left translation upon each other.
+"""
+function compose(G::AbstractGroupManifold, x...)
+    return reduce((x, y)->left_action(G, x, y), x)
+end
+
+*(G::AbstractGroupManifold, x, y) = compose(G, x, y)
+
+/(G::AbstractGroupManifold, x, y) = compose(G, x, inv(G, y))
+
+\(G::AbstractGroupManifold, x, y) = compose(G, inv(G, x), y)
+
+@doc doc"""
+    conjugate(G::AbstractGroupManifold, x, y)
+
+Compute the conjugate of the element $y$ by element $x$, i.e.
+$x \cdot y \cdot x^{-1}$
+"""
+function conjugate(G::AbstractGroupManifold, x, y)
+    return compose(G, x, y, inv(G, x))
+end
+
+function adjoint_representation(G::AbstractGroupManifold, x, v)
+    return left_action(G, left_action(G, x, v), inv(G, x))
+end
+
+function adjoint_representation_derivative(G::AbstractGroupManifold, v, w)
+    return lie_bracket(G, v, w)
+end
+
+function lie_bracket(G::AbstractGroupManifold, v, w)
+    error("lie_bracket not implemented on $(typeof(G)) for vectors $(typeof(v)) and $(typeof(w))")
+end
+
+function inner(G::AbstractGroupManifold, x, v, w)
+    return dot(left_action(G, x, v), left_action(G, x, w))
```

Comment:
Is there any example where this gives different result from `dot(v, w)`?

Reply:
Since the left action brings you into the Lie algebra, there might be a more efficient way in there, I think; though I do not have an example at hand.